README.md

```diff
@@ -1,7 +1,214 @@
+        // adjust the request here using request.copyWith
+        // Maybe add auth headers?
+        // add additional content to the body?
+    
+        // do the call that returns the response
+        final response = await handler(request);
+      
+        // do additional response handling
+        // maybe resend if the response failed because of an expired token?
+        // maybe log out the user if he has no permission to this endpoint?
+        return response;
+      };
+}
+```
+
+An example could look like this:
+
+```dart
+class AuthInterceptor implements Interceptor {
+  @override
+  RequestHandler<T> intercept<T>(RequestHandler<T> handler) => (request) async {
+        final updatedRequest = request.copyWith(
+          headers: {"Authorization": await getToken()},
+        );
+
+	return handler(updatedRequest);
+  };
+}
+```
+
+or
+
+```dart
+class RefreshTokenInterceptor implements Interceptor {
+  @override
+  RequestHandler<T> intercept<T>(RequestHandler<T> handler) => (request) async {
+        final response = await handler(updatedRequest);
+
+        if (!response.isSuccessful && response.errorBody == "Shit! Token expired!") {
+          // refresh the token asynchronously
+          final refreshedRequest = request.copyWith(
+            headers: {"Authorization": await refreshToken()},
+          );
+          
+          // resend the request with the refreshed token!
+          return handler(refreshedRequest);
+        }
+
+        return response;
+  };
+}
+```
+
+
+
+2. #### Chain your interceptors
+
+   You can chain multiple `Interceptor`'s using `Interceptor.fromList`. 
+
+   ```dart
+   final apiProvider = HttpClientApiProvider(
+       httpClient: yourClient,
+       container: yourContainer,
+       interceptor: Interceptor.fromList([
+         TokenInterceptor(),
+         MyInterceptor(),
+         LoggingInterceptor(),
+       ]),
+   )
+   ```
+
+   The given `Interceptors`'s wrap around the source call like an onion.  🤤
+
+   TODO insert graphic
+
+   
+
+   Testing a chain of `Interceptor`'s can be done by mocking.
+
+   
+
+   
+
+   
+
+   ### Testing / Mocking an `ApiProvider`
+
+   The slim interface of `ApiProvider` makes mocking really easy. You can simply implement the `ApiProvider` interface, implement the `request` method and returns whatever you need in your use case. 
+
+   
+
+   Besides that we provide a powerful mocking implementation out of the Box!
+
+   Use `MockApiProvider` and configure it using the builder!
+
+   ```dart
+   final apiProvider = MockApiProvider((builder) {
+       builder.onGet<User>(
+         url: Path("https://someapi/user"),
+         handler: (request) async => Response.success(body: mockedUser()),
+       );
+   });
+   ```
+
+   If you want to test your `ApiProvider` with all of your custom `Interceptor`'s' we gotcha covered, because `MockApiProvider` supports  `Interceptor`'s   as well! 💪🏼 
```

Comment:
Another exclamation mark spotted here. Do you see what I mean?

Reply:
ja!!! 😠 removed it! ✅